### PR TITLE
sanitization strategy now has defaults

### LINKED ISSUE #270: System.ArgumentException: RuleName
Getting the following exception:

> System.ArgumentException: RuleName 'GatedPublishRunner+Event' contains character(s) that are not allowed by Service Bus. Entity names or path can contain only letters, numbers, periods (.), hyphens (-), and underscores (_).

Based on the usage of **ValidateAndHashIfNeeded** I would not expect this exception.

Using the *V6Sanitization* snippet does not result in this exception likely to indicate that *ValidateAndHashIfNeeded* is broken.

The last NServiceBus code in the stacktrace is:

     at NServiceBus.AzureServiceBus.AzureServiceBusForwardingSubscriptionCreator.<Create>d__5.MoveNext()

Maybe the *sanitization* is not applied everywhere or maybe the sanitization is incorrect.

This is the initialization:

```c#
        endpointConfiguration
            .UseTransport<AzureServiceBusTransport>()
            .UseTopology<ForwardingTopology>()
            .ConnectionString(connectionstring)
            .Sanitization().UseStrategy<ValidateAndHashIfNeeded>();
```

## Packages

```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="NServiceBus" version="6.0.0-beta0004" targetFramework="net452" />
  <package id="NServiceBus.Azure.Transports.WindowsAzureServiceBus" version="7.0.0-beta0007" targetFramework="net452" />
  <package id="WindowsAzure.ServiceBus" version="3.3.0" targetFramework="net452" />
</packages>
``` 

## Stack trace
```
|FATAL|1|Host.Program|Main
System.ArgumentException: RuleName 'GatedPublishRunner+Event' contains character(s) that are not allowed by Service Bus. Entity names or path can contain only letters, numbers, periods (.), hyphens (-), and underscores (_).
Parameter name: RuleName TrackingId:3776a263-ce2c-4166-a4ac-ddfdfadc77a7_G23_G23_B28, SystemTracker:ramon-test:Topic:bundle-1|PerformanceTest, Timestamp:6/28/2016 9:56:50 AM ---> System.ServiceModel.FaultException`1[System.ServiceModel.ExceptionDetail]: RuleName 'GatedPublishRunner+Event' contains character(s) that are not allowed by Service Bus. Entity names or path can contain only letters, numbers, periods (.), hyphens (-), and underscores (_).
Parameter name: RuleName TrackingId:3776a263-ce2c-4166-a4ac-ddfdfadc77a7_G23_G23_B28, SystemTracker:ramon-test:Topic:bundle-1|PerformanceTest, Timestamp:6/28/2016 9:56:50 AM

Server stack trace: 
   at Microsoft.ServiceBus.Messaging.Sbmp.DuplexRequestBindingElement.DuplexRequestSessionChannel.ThrowIfFaultMessage(Message wcfMessage)
   at Microsoft.ServiceBus.Messaging.Sbmp.DuplexRequestBindingElement.DuplexRequestSessionChannel.HandleMessageReceived(IAsyncResult result)

Exception rethrown at [0]: 
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Sbmp.DuplexRequestBindingElement.DuplexRequestSessionChannel.EndRequest(IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Channels.ReconnectBindingElement.ReconnectChannelFactory`1.RequestSessionChannel.RequestAsyncResult.<>c.<GetAsyncSteps>b__9_3(RequestAsyncResult thisPtr, IAsyncResult r)
   at Microsoft.ServiceBus.Messaging.IteratorAsyncResult`1.StepCallback(IAsyncResult result)

Exception rethrown at [1]: 
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Channels.ReconnectBindingElement.ReconnectChannelFactory`1.RequestSessionChannel.EndRequest(IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Sbmp.RedirectBindingElement.RedirectContainerChannelFactory`1.RedirectContainerSessionChannel.RequestAsyncResult.<>c__DisplayClass8_1.<GetAsyncSteps>b__4(RequestAsyncResult thisPtr, IAsyncResult r)
   at Microsoft.ServiceBus.Messaging.IteratorAsyncResult`1.StepCallback(IAsyncResult result)

Exception rethrown at [2]: 
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Sbmp.RedirectBindingElement.RedirectContainerChannelFactory`1.RedirectContainerSessionChannel.EndRequest(IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Channels.ReconnectBindingElement.ReconnectChannelFactory`1.RequestSessionChannel.RequestAsyncResult.<>c.<GetAsyncSteps>b__9_3(RequestAsyncResult thisPtr, IAsyncResult r)
   at Microsoft.ServiceBus.Messaging.IteratorAsyncResult`1.StepCallback(IAsyncResult result)

Exception rethrown at [3]: 
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Channels.ReconnectBindingElement.ReconnectChannelFactory`1.RequestSessionChannel.EndRequest(IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Sbmp.SbmpTransactionalAsyncResult`1.<>c.<GetAsyncSteps>b__18_3(TIteratorAsyncResult thisPtr, IAsyncResult a)
   at Microsoft.ServiceBus.Messaging.IteratorAsyncResult`1.StepCallback(IAsyncResult result)

Exception rethrown at [4]: 
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.Sbmp.SbmpSubscriptionClient.OnEndAddRule(IAsyncResult result)
   --- End of inner exception stack trace ---

Server stack trace: 
   at Microsoft.ServiceBus.Messaging.Sbmp.SbmpSubscriptionClient.OnEndAddRule(IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.OpenOnceManager.<>c__DisplayClass6_0.<Begin>b__0(IAsyncResult r, Boolean& output)
   at Microsoft.ServiceBus.Messaging.OpenOnceManager.OpenOnceManagerAsyncResult`1.OperationComplete(IAsyncResult result)
   at Microsoft.ServiceBus.Common.AsyncResult.AsyncCompletionWrapperCallback(IAsyncResult result)

Exception rethrown at [0]: 
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.SubscriptionClient.EndAddRule(IAsyncResult result)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NServiceBus.AzureServiceBus.AzureServiceBusForwardingSubscriptionCreator.<Create>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.AzureServiceBus.TopologyCreator.<Create>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.AzureServiceBus.SubscriptionManager.<Subscribe>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Features.AutoSubscribe.ApplySubscriptions.<OnStart>d__1.MoveNext() in C:\Build\src\NServiceBus.Core\Routing\AutomaticSubscriptions\AutoSubscribe.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Features.FeatureActivator.<StartFeatures>d__5.MoveNext() in C:\Build\src\NServiceBus.Core\Features\FeatureActivator.cs:line 71
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.StartableEndpoint.<StartFeatures>d__4.MoveNext() in C:\Build\src\NServiceBus.Core\StartableEndpoint.cs:line 84
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.StartableEndpoint.<Start>d__1.MoveNext() in C:\Build\src\NServiceBus.Core\StartableEndpoint.cs:line 44
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Endpoint.<Start>d__1.MoveNext() in C:\Build\src\NServiceBus.Core\Endpoint.cs:line 28
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at BaseRunner.<CreateEndpoint>d__27.MoveNext() in S:\particular\EndToEnd\src\PerformanceTests\Common\Scenarios\BaseRunner.cs:line 230
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at BaseRunner.<Execute>d__21.MoveNext() in S:\particular\EndToEnd\src\PerformanceTests\Common\Scenarios\BaseRunner.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Host.Program.Main() in S:\particular\EndToEnd\src\PerformanceTests\Common\Program.cs:line 49
```

### PR DESCRIPTION
**Pull request by yvesgoeleven**
Fixes #270

### CODE REVIEW

File: src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs [hunk 1]
```diff
@@ -1,17 +1,116 @@
...
     {
         readonly ReadOnlySettings settings;
 
+        // Entity segments can contain only letters, numbers, periods (.), hyphens (-), and underscores (-), paths can contain slashes (/)
+        Regex queueAndTopicNameRegex = new Regex(@"^[^\/][0-9A-Za-z_\.\-\/]+[^\/]$");
+
+        // Except for subscriptions and rules, these cannot contain slashes (/)
+        Regex subscriptionAndRuleNameRegex = new Regex(@"^[0-9A-Za-z_\.\-]+$");
+
+        Regex sanitizationRegex = new Regex(@"[^a-zA-Z0-9\-\._]");
```
Comment (yvesgoeleven) [inline src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs]: V6 one, change to V7

Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs]: Documentation on [`ValidateAndHashIfNeeded`](http://docs.particular.net/nservicebus/azure-service-bus/sanitization#automated-sanitization-versions-7-and-above) will need to be updated @yvesgoeleven

**Comment by SeanFeldman** [comment]
@Particular/azure-service-bus-maintainers issue with sanitization rules resolved. Please review.

File: src/Tests/API/APIApprovals.ApproveAzureServiceBusTransport.approved.txt [hunk 2]
```diff
@@ -48,6 +48,7 @@
     public interface ISanitizationStrategy
     {
         string Sanitize(string entityPathOrName, NServiceBus.AzureServiceBus.EntityType entityType);
+        void SetDefaultRules(NServiceBus.Settings.SettingsHolder settings);
```
Comment (melkio) [inline src/Tests/API/APIApprovals.ApproveAzureServiceBusTransport.approved.txt]: I don't like this "new design". It's not a _sanitization strategy_ concern setting up default values. What a _sanitization strategy_ has to do is to sanitize.
Moreover, if a user wants to implement a new strategy, they need to know what to do to setup default rules and then go through the transport code to learn which `setting keys` have to configure

**Comment by melkio** [comment]
To make the `ISanitizationStrategy` design simpler, we could remove `SetDefaultRules` method and instead of retrieving validation/sanitization func with `settings.GetOrDefault<...>(...)` method we could use `settings.TryGet<...>(...)`. 
If users have provided a custom func this method will return "something" otherwise _sanitization strategy_ can use its own

@Particular/azure-service-bus-maintainers thoughts?

**Comment by yvesgoeleven** [comment]
@melkio good point... I'll undo the api change

**Comment by SeanFeldman** [comment]
Added a few minor tweaks.

**Comment by melkio** [comment]
What about splitting `sanitization strategy` into two components? Something like:

```csharp

public interface ISanitizationStrategy
{
    Func<string, ValidationResult> QueueValidator;
    Func<string, string> QueueSanitaizer;
    // the same for others entity types
}

class CustomSanitizationStrategy : ISanitizationStrategy
{
    public Func<string, ValidationResult> QueueValidator = x => ...
    public Func<string, string> QueueSanitaizer = x => ...
    /// ...
}

class Sanitizer
{
    public Sanitizer(ReadOnlySettings settings)
    {
        // ...
    }

    public string Sanitize(string entityPathOrName, EntityType entityType)
    {
        Func<string, ValidationResult> validator;
        Func<string, string> sanitizer;

        switch (entityType)
        {
            case EntityType.Queue:
                validator = settings.GetOrDefault<Func<string, ValidationResult>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.QueuePathValidator);
                sanitizer = settings.GetOrDefault<Func<string, string>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.QueuePathSanitizer);
                break;

            case EntityType.Topic:
                validator = settings.GetOrDefault<Func<string, ValidationResult>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.TopicPathValidator);
                sanitizer = settings.GetOrDefault<Func<string, string>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.TopicPathSanitizer);
                break;

            case EntityType.Subscription:
                validator = settings.GetOrDefault<Func<string, ValidationResult>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.SubscriptionNameValidator);
                sanitizer = settings.GetOrDefault<Func<string, string>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.SubscriptionNameSanitizer);
                break;

            case EntityType.Rule:
                validator = settings.GetOrDefault<Func<string, ValidationResult>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.RuleNameValidator);
                sanitizer = settings.GetOrDefault<Func<string, string>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.RuleNameSanitizer);
                break;

            default:
                throw new ArgumentOutOfRangeException(nameof(entityType), entityType, null);
        }

        var validationResult = validator(entityPathOrName);
        if (validationResult.IsValid)
            return entityPathOrName;

        var sanitizedValue = entityPathOrName;
        if (!validationResult.CharactersAreValid)
            sanitizedValue = sanitizer(entityPathOrName);

        // second validation pass to validate length based on sanitized characters
        validationResult = validator(sanitizedValue);
        if (validationResult.LengthIsValid)
            return sanitizedValue;

        return settings.GetOrDefault<Func<string, string>>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.Hash)(sanitizedValue);        
    }
}


```

**Comment by yvesgoeleven** [comment]
@melkio this skeleton won't work for the Throw scenario

**Comment by melkio** [comment]
Yes. It could work. 
`ValidationFunction(s)` can throw if provided value isn't fine

**Comment by yvesgoeleven** [comment]
@melkio but then you need different validations methods again, bringing back the original issue back again

**Comment by melkio** [comment]
What do you mean with:

> you need different validations methods again

`ThrowOnFailedValidation` and `ValidateAndHashIfNeeded` provide one different default function to validate/sanitize each `entity type`
What I'm suggesting is to expose these functions from the interface contract, to be more expressive if a user wants to implements their own strategy

**Comment by melkio** [comment]
Since this PR is fine by me, we can merge it, and talk about improvements or related things later.

**Pull request merged**